libraries/botbuilder-core/botbuilder/core/teams/teams_activity_handler.py

```diff
@@ -0,0 +1,404 @@
+    ):
+        return
+
+    async def on_teams_team_renamed_activity(  # pylint: disable=unused-argument
+        self, team_info: TeamInfo, turn_context: TurnContext
+    ):
+        return
+
+    async def on_teams_members_added_dispatch_activity(  # pylint: disable=unused-argument
+        self,
+        members_added: [ChannelAccount],
+        team_info: TeamInfo,
+        turn_context: TurnContext,
+    ):
+        """
+        team_members = {}
+        team_members_added = []
+        for member in members_added:
+            if member.additional_properties != {}:
+                team_members_added.append(TeamsChannelAccount(member))
+            else:
+                if team_members == {}:
+                    result = await TeamsInfo.get_members_async(turn_context)
+                    team_members = { i.id : i for i in result }
+
+                if member.id in team_members:
+                    team_members_added.append(member)
+                else:
+                    newTeamsChannelAccount = TeamsChannelAccount(
+                        id=member.id,
+                        name = member.name,
+                        aad_object_id = member.aad_object_id,
+                        role = member.role
+                        )
+                    team_members_added.append(newTeamsChannelAccount)
+
+        return await self.on_teams_members_added_activity(teams_members_added, team_info, turn_context)
+        """
+        for member in members_added:
+            new_account_json = member.seralize()
+            del new_account_json["additional_properties"]
+            member = TeamsChannelAccount(**new_account_json)
+        return await self.on_teams_members_added_activity(members_added, turn_context)
+
+    async def on_teams_members_added_activity(
+        self, teams_members_added: [TeamsChannelAccount], turn_context: TurnContext
+    ):
+        for member in teams_members_added:
+            member = ChannelAccount(member)
+        return super().on_members_added_activity(teams_members_added, turn_context)
+
+    async def on_teams_members_removed_dispatch_activity(  # pylint: disable=unused-argument
+        self,
+        members_removed: [ChannelAccount],
+        team_info: TeamInfo,
+        turn_context: TurnContext,
+    ):
+        teams_members_removed = []
+        for member in members_removed:
+            new_account_json = member.seralize()
+            del new_account_json["additional_properties"]
+            teams_members_removed.append(TeamsChannelAccount(**new_account_json))
+
+        return await self.on_teams_members_removed_activity(
+            teams_members_removed, turn_context
+        )
+
+    async def on_teams_members_removed_activity(
+        self, teams_members_removed: [TeamsChannelAccount], turn_context: TurnContext
+    ):
+        members_removed = [ChannelAccount(i) for i in teams_members_removed]
```

Comment:
Is the cast needed?  TeamsChannelAccount is a ChannelAccount already.

Reply:
It should be needed, TeamsChannelAccount is a richer ChannelAccount. Since we are calling the base on_members_removed which takes a list of ChannelAccount to make it happy.

Reply:
Technically it doesn't hurt anything.  But a cast of a TeamsChannelAccount to a ChannelAccount is still a TeamsChannelAccount.  If you were to compare the contents of the list, you'd see they were the same objects:  A TeamsChannelAccount.  C# has to cast because of strong typing.

What's also interesting is ActivityHandler.on_members_added_activity is declared as wanting a ChannelAccount, not a list.  Python won't care, it will run the same.  But in C# it is declared as requiring a List.

As a side note, ActivityHandler uses the "_activity" suffix on those methods.  We need to coordinate with Axel on this since John requested the "_activity" be removed on the TeamsActivityHandler.  May not be a change we want to make at this point.